cookbook/case_studies/ml_training/mnist_classifier/pytorch_single_node_multi_gpu.py

```diff
@@ -0,0 +1,406 @@
+    log_interval: int = 10
+    batch_size: int = 64
+    test_batch_size: int = 1000
+    epochs: int = 10
+    learning_rate: float = 0.01
+
+
+# %%
+# Training and Evaluating
+# =======================
+
+TrainingOutputs = typing.NamedTuple(
+    "TrainingOutputs",
+    epoch_accuracies=typing.List[float],
+    model_state=PythonPickledFile,
+)
+
+MODEL_FILE = "./mnist_cnn.pt"
+ACCURACIES_FILE = "./mnist_cnn_accuracies.json"
+
+
+def dist_setup(rank, world_size, backend):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "8888"
+    dist.init_process_group(backend, rank=rank, world_size=world_size)
+
+
+def train_mnist(rank: int, world_size: int, hp: Hyperparameters) -> TrainingOutputs:
+
+    # store the hyperparameters' config in ``wandb``
+    if rank == 0:
+        wandb_setup()
+        wandb.config.update(json.loads(hp.to_json()))
+        print("wandb config:", wandb.config)
+
+    # set random seed
+    torch.manual_seed(hp.seed)
+
+    use_cuda = torch.cuda.is_available()
+    print(f"Using distributed PyTorch with {hp.backend} backend")
+    print(f"Running MNIST training on rank {rank}, world size: {world_size}")
+    print(f"Use cuda: {use_cuda}")
+    dist_setup(rank, world_size, hp.backend)
+    print(f"Rank {rank + 1}/{world_size} process initialized.\n")
+
+    # load data
+    kwargs = {"num_workers": 0, "pin_memory": True} if use_cuda else {}
+    print("Getting data loaders")
+    training_data_loader = mnist_dataloader(
+        DATA_DIR, hp.batch_size, train=True, distributed=use_cuda, rank=rank, world_size=world_size, **kwargs
+    )
+    test_data_loader = mnist_dataloader(DATA_DIR, hp.test_batch_size, train=False, **kwargs)
+
+    # define the distributed model and optimizer
+    print("Defining model")
+    model = Net().cuda(rank)
+    model = nn.parallel.DistributedDataParallel(model, device_ids=[rank])
+
+    optimizer = optim.SGD(model.parameters(), lr=hp.learning_rate, momentum=hp.sgd_momentum)
+
+    # train the model: run multiple epochs and capture the accuracies for each epoch
+    print(f"Training for {hp.epochs} epochs")
+    accuracies = []
+    for epoch in range(1, hp.epochs + 1):
+        train(model, rank, training_data_loader, optimizer, epoch, hp.log_interval)
+
+        # only compute validation metrics in the main process
+        if rank == 0:
+            accuracies.append(test(model, rank, test_data_loader))
+        dist.barrier()  # wait for main process to complete validation before continuing training
+
+    if rank == 0:
+        wandb.finish()  # this is important to tell to wandb that we're done logging metrics
+
+        # after training the model, we can simply save it to disk and return it from the Flyte task as a :py:class:`flytekit.types.file.FlyteFile`
+        # type, which is the ``PythonPickledFile``. ``PythonPickledFile`` is simply a decorator on the ``FlyteFile`` that records the format
+        # of the serialized model as ``pickled``
```

Comment:
I think we can remove this as we aren't returning the `MODEL_FILE`. Moreover, this isn't a Flyte task.

Reply:
which function are you referring to?

Reply:
oh do you mean `pytorch_mnist_task`? I don't think we can, since we `mp.spawn` doesn't actually return anything. We'd need to use a `Queue` object to serialize the model and I think that adds too many complications to this tutorial, which is why it's simpler to save to disk in the main (rank == 0) process

Reply:
Okay.